File: Problems/Algorithms/200. Number of Islands/number_of_islands.go
func numIslands(grid [][]byte) int {
    n := len(grid)
    m := len(grid[0])

    ans := 0
    for i := 0; i < n; i++ {
        for j := 0; j < m; j++ {
            ans += dfs(grid, n, m, i, j)
        }
    }

    return ans
}

func isValid(n int, m int, i int, j int) bool {
    return i >= 0 && i < n && j >= 0 && j < m
}

func dfs(grid [][]byte, n int, m int, i int, j int) int {
    if !isValid(n, m, i, j) || grid[i][j] == '0' {
        return 0
    }

    grid[i][j] = '0'

    dfs(grid, n, m, i+1, j)
    dfs(grid, n, m, i-1, j)
    dfs(grid, n, m, i, j+1)
    dfs(grid, n, m, i, j-1)

    return 1
}

File: Problems/Algorithms/200. Number of Islands/number_of_islands.rs
impl Solution {
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let mut ans = 0;
        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                ans += Self::dfs(&mut grid, i, j);
            }
        }
    
        ans
    }

    fn dfs(grid: &mut Vec<Vec<char>>, i: usize, j: usize) -> i32 {
        if i < 0 || i >= grid.len() || j < 0 || j >= grid[0].len() || grid[i][j] == '0' {
            return 0;
        }
        
        grid[i][j] = '0';
        Self::dfs(grid, i-1, j);
        Self::dfs(grid, i+1, j);
        Self::dfs(grid, i, j-1);
        Self::dfs(grid, i, j+1);
        
        1
    }
}
